calico/felix/ipsets.py

```diff
@@ -544,12 +544,16 @@ def _finish_msg_batch(self, batch, results):
             self._force_reprogram = False
 
     def _sync_to_ipset(self):
-        _log.info("Rewriting %s with %d members.", self, len(self.members))
         _log.debug("Setting ipset %s to %s", self, self.members)
-        # Defer to our helper.
-        self._ipset.replace_members(self.members)
-        # We have got the set into the correct state.
-        self.programmed_members = self.members.copy()
+        # Defer to our helper to actually make the changes.
+        if self._force_reprogram or self.programmed_members is None:
+            self._ipset.replace_members(self.members)
+            self.programmed_members = self.members.copy()
```

Comment:
Since we're on an occupancy push, why don't we rework the code a little to avoid these copies?  We could adjust the contract of replace_members to require that the input is a copy (it already is, in fact), then directly store it as self.members rather than doing a clear/update; then directly copy it to self.programmed_members.

I wish Python had an immutable set type, which would make it much harder to shoot ourselves in the foot when passing sets around between actors.